pkg/registry/registry.go

```diff
@@ -0,0 +1,211 @@
+package registry
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/suse-edge/edge-image-builder/pkg/fileio"
+	"github.com/suse-edge/edge-image-builder/pkg/http"
+	"github.com/suse-edge/edge-image-builder/pkg/image"
+	"go.uber.org/zap"
+)
+
+type helmClient interface {
+	AddRepo(repository *image.HelmRepository) error
+	RegistryLogin(repository *image.HelmRepository) error
+	Pull(chart string, repository *image.HelmRepository, version, destDir string) (string, error)
+	Template(chart, repository, version, valuesFilePath, kubeVersion, targetNamespace string) ([]map[string]any, error)
+}
+
+type helmChart struct {
+	image.HelmChart
+	localPath     string
+	repositoryURL string
+}
+
+type Registry struct {
+	embeddedImages []image.ContainerImage
+	manifestsDir   string
+	helmClient     helmClient
+	helmCharts     []*helmChart
+	helmValuesDir  string
+	kubeVersion    string
+}
+
+func New(ctx *image.Context, localManifestsDir string, helmClient helmClient, helmValuesDir string) (*Registry, error) {
+	manifestsDir, err := storeManifests(ctx, localManifestsDir)
+	if err != nil {
+		return nil, fmt.Errorf("storing manifests: %w", err)
+	}
+
+	charts, err := storeHelmCharts(ctx, helmClient)
```

Comment:
This is an interesting philosophical discussion. I was always taught to not to heavyweight operations in a constructor by default. Granted, this is more of a factory than a constructor, and Go has plenty of other differences from Java & Python that I'm still adjusting to. So I'm certainly not pushing you to change it, but if you were at all on the fence that this felt like the wrong place to do it, then you're not alone.

Reply:
I'd want to see how it turns out. I'm not quite happy with it exactly due to the reasons you listed, but at the same time I need to do this first because the combustion component runs.

The alternative would be for the constructor  to only initialise some of the struct fields and have another function e.g. `Prepare()` to do the downloading right after, however I find this even more confusing.

Reply:
I've had the same debate in my own head, where I've split the constructor and an init() function. The benefits were there, but minor. In any event, I'm totally cool with your rationale.